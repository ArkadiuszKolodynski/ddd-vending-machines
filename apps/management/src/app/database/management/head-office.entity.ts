import { BaseEntity, Embedded, Entity, PrimaryKey, Property } from '@mikro-orm/core';
import { MoneyEmbeddable } from '@vending-machines/shared';

@Entity({ tableName: 'head_office' })
export default class HeadOfficeEntity extends BaseEntity<HeadOfficeEntity, 'id'> {
  @PrimaryKey({ type: 'uuid' })
  id!: string;

  @Property({ type: 'decimal', precision: 18, scale: 2 })
  balance!: number;

  @Embedded(() => MoneyEmbeddable)
  cash!: MoneyEmbeddable;
}
