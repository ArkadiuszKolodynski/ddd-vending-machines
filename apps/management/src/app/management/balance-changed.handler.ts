import { Logger } from '@nestjs/common';
import { EventsHandler, IEventHandler } from '@nestjs/cqrs';
import { BalanceChangedEvent } from '@vending-machines/events';
import Currency from 'currency.js';
import { HeadOfficeRepository } from './head-office.repository.interface';

@EventsHandler(BalanceChangedEvent)
export class BalanceChangedHandler implements IEventHandler<BalanceChangedEvent> {
  constructor(private readonly headOfficeRepository: HeadOfficeRepository) {}

  async handle(event: BalanceChangedEvent) {
    const headOffice = await this.headOfficeRepository.findOne(event.aggregateId);

    if (!headOffice) {
      Logger.error(`Head office with id ${event.aggregateId} not found`, this.constructor.name);
      return;
    }

    headOffice.changeBalance(new Currency(event.payload.delta));
    await this.headOfficeRepository.save(headOffice);
  }
}
