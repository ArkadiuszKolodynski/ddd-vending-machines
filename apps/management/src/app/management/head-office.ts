import { AggregateRoot, Money } from '@vending-machines/shared';
import Currency from 'currency.js';

export class HeadOffice extends AggregateRoot {
  balance: Currency = new Currency(0);
  cash: Money = Money.None;

  changeBalance(delta: Currency): void {
    this.balance = this.balance.add(delta);
  }
}
