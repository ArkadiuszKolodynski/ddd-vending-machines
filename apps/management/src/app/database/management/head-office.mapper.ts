import { RequiredEntityData } from '@mikro-orm/core';
import { Money } from '@vending-machines/shared';
import Currency from 'currency.js';
import { HeadOffice } from '../../management/head-office';
import HeadOfficeEntity from './head-office.entity';

export class HeadOfficeMapper {
  private constructor() {}

  static toDomain(entity: HeadOfficeEntity): HeadOffice {
    const headOffice = new HeadOffice();
    return Object.assign(headOffice, {
      id: entity.id,
      balance: new Currency(entity.balance),
      cash: new Money(
        entity.cash.oneCentCount,
        entity.cash.tenCentCount,
        entity.cash.quarterCount,
        entity.cash.oneDollarCount,
        entity.cash.fiveDollarCount,
        entity.cash.twentyDollarCount,
      ),
    });
  }

  static toPersistence(headOffice: HeadOffice): RequiredEntityData<HeadOfficeEntity> {
    return {
      id: headOffice.id,
      balance: headOffice.balance.value,
      cash: { ...headOffice.cash },
    };
  }
}
