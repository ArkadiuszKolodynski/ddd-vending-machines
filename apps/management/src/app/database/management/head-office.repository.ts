import { EntityManager, EntityRepository } from '@mikro-orm/core';
import { InjectRepository } from '@mikro-orm/nestjs';
import { Injectable, Provider } from '@nestjs/common';
import { HeadOffice } from '../../management/head-office';
import { HeadOfficeRepository } from '../../management/head-office.repository.interface';
import HeadOfficeEntity from './head-office.entity';
import { HeadOfficeMapper } from './head-office.mapper';

@Injectable()
export class MikroOrmHeadOfficeRepository implements HeadOfficeRepository {
  constructor(
    @InjectRepository(HeadOfficeEntity)
    private readonly headOfficeRepository: EntityRepository<HeadOfficeEntity>,
    private readonly em: EntityManager,
  ) {}

  async findOne(id: string): Promise<HeadOffice | null> {
    const headOfficeEntity = await this.headOfficeRepository.findOne({ id });

    if (!headOfficeEntity) {
      return null;
    }

    return HeadOfficeMapper.toDomain(headOfficeEntity);
  }

  async save(headOffice: HeadOffice): Promise<void> {
    let headOfficeEntity = await this.headOfficeRepository.findOne({ id: headOffice.id });

    if (headOfficeEntity) {
      headOfficeEntity.assign(HeadOfficeMapper.toPersistence(headOffice));
    } else {
      headOfficeEntity = this.headOfficeRepository.create(HeadOfficeMapper.toPersistence(headOffice));
    }

    await this.em.flush();
  }
}

export const HeadOfficeRepositoryProvider: Provider = {
  provide: HeadOfficeRepository,
  useClass: MikroOrmHeadOfficeRepository,
};
