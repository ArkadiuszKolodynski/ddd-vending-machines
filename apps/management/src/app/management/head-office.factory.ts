import { Injectable } from '@nestjs/common';
import { EventPublisher } from '@nestjs/cqrs';
import { HeadOfficeCreatedEvent } from '@vending-machines/events';
import { HeadOffice } from './head-office';

@Injectable()
export class HeadOfficeFactory {
  constructor(private readonly eventPublisher: EventPublisher) {}

  create(id?: string): HeadOffice {
    const headOffice = this.eventPublisher.mergeObjectContext(new HeadOffice(id));
    headOffice.apply(
      new HeadOfficeCreatedEvent({ aggregateId: id, aggregateType: headOffice.constructor.name, payload: {} }),
    );
    return headOffice;
  }
}
