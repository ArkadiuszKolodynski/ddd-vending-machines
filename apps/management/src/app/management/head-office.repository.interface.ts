import { HeadOffice } from './head-office';

export interface HeadOfficeRepository {
  findOne(id: string): HeadOffice | null | Promise<HeadOffice | null>;
  save(headOffice: HeadOffice): void | Promise<void>;
}

export const HeadOfficeRepository = Symbol('HeadOfficeRepository');
