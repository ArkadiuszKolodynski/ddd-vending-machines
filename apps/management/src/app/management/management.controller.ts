import { Controller, Get, HttpCode, HttpStatus, Param, ParseUUIDPipe, Post } from '@nestjs/common';
import { CommandBus, QueryBus } from '@nestjs/cqrs';
import { LoadCashToAtmCommand } from './commands/impl/load-cash-to-atm.command';
import { UnloadCashFromSnackMachineCommand } from './commands/impl/unload-cash-from-snack-machine.command';
import { HeadOfficeDto } from './dto/head-office.dto';
import { GetHeadOfficeQuery } from './queries/impl/get-head-office.query';

@Controller('management')
export class ManagementController {
  constructor(
    private readonly commandBus: CommandBus,
    private readonly queryBus: QueryBus,
  ) {}

  @Get(':id')
  getHeadOffice(@Param('id', new ParseUUIDPipe({ version: '4' })) id: string): Promise<HeadOfficeDto> {
    return this.queryBus.execute(new GetHeadOfficeQuery(id));
  }

  @Post(':id/load-cash-to-atm')
  @HttpCode(HttpStatus.OK)
  async loadCashToAtm(@Param('id', new ParseUUIDPipe({ version: '4' })) id: string): Promise<HeadOfficeDto> {
    await this.commandBus.execute(new LoadCashToAtmCommand(id));
    return this.queryBus.execute(new GetHeadOfficeQuery(id));
  }

  @Post(':id/unload-cash-from-snack-machine')
  @HttpCode(HttpStatus.OK)
  async unloadCashFromSnackMachine(
    @Param('id', new ParseUUIDPipe({ version: '4' })) id: string,
  ): Promise<HeadOfficeDto> {
    await this.commandBus.execute(new UnloadCashFromSnackMachineCommand(id));
    return this.queryBus.execute(new GetHeadOfficeQuery(id));
  }
}
