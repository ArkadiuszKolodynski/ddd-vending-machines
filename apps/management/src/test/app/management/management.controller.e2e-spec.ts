import { HttpStatus, INestApplication, Module } from '@nestjs/common';
import { CommandBus, QueryBus } from '@nestjs/cqrs';
import { Test } from '@nestjs/testing';
import request from 'supertest';
import { HeadOfficeRepository } from '../../../app/management/head-office.repository.interface';

@Module({
  providers: [{ provide: HeadOfficeRepository, useValue: { findOne: jest.fn(), save: jest.fn() } }],
  exports: [HeadOfficeRepository],
})
class DatabaseModuleMock {}
jest.mock('../../../app/database/database.module', () => {
  return {
    DatabaseModule: DatabaseModuleMock,
  };
});

import { ValidationProvider } from '@vending-machines/shared';
import { ManagementController } from '../../../app/management/management.controller';

describe('ManagementController - e2e', () => {
  const queryBusMock = { execute: () => ({ balance: '$1.00', cash: '$1.00' }) };
  const testEndpoint = '/management';

  let app: INestApplication;

  beforeAll(async () => {
    const module = await Test.createTestingModule({
      controllers: [ManagementController],
      providers: [
        ValidationProvider,
        { provide: QueryBus, useValue: queryBusMock },
        { provide: CommandBus, useValue: { execute: jest.fn() } },
      ],
    }).compile();

    app = module.createNestApplication();
    await app.init();
  });

  afterAll(async () => {
    await app.close();
  });

  describe('GET /', () => {
    it('should return 200 OK', () => {
      return request(app.getHttpServer()).get(testEndpoint).expect(HttpStatus.OK).expect(queryBusMock.execute());
    });
  });
});
