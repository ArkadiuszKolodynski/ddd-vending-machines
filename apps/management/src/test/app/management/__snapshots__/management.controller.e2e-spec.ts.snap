// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManagementController - e2e GET / should return 400 BAD REQUEST when id is not a valid uuid 1`] = `"Validation failed (uuid v 4 is expected)"`;
